satellite/metabase/rangedloop/service_test.go

```diff
@@ -149,3 +151,202 @@ func TestLoopCancellation(t *testing.T) {
+				incShouldBeCalled()
+				return nil, nil
+			},
+			OnProcess: func(ctx context.Context, segments []segmentloop.Segment) error {
+				incShouldBeCalled()
+				return errors.New("Test OnProcess error")
+			},
+			OnJoin: func(ctx context.Context, partial rangedloop.Partial) error {
+				require.Fail(t, "OnJoin should not be called")
+				return nil
+			},
+			OnFinish: func(ctx context.Context) error {
+				require.Fail(t, "OnFinish should not be called")
+				return nil
+			},
+		},
+		&rangedlooptest.CallbackObserver{
+			OnStart: func(ctx context.Context, t time.Time) error {
+				incShouldBeCalled()
+				return nil
+			},
+			OnFork: func(ctx context.Context) (rangedloop.Partial, error) {
+				incShouldBeCalled()
+				return nil, nil
+			},
+			OnProcess: func(ctx context.Context, segments []segmentloop.Segment) error {
+				incShouldBeCalled()
+				return nil
+			},
+			OnJoin: func(ctx context.Context, partial rangedloop.Partial) error {
+				incShouldBeCalled()
+				return errors.New("Test OnJoin error")
+			},
+			OnFinish: func(ctx context.Context) error {
+				require.Fail(t, "OnFinish should not be called")
+				return nil
+			},
+		},
+		&rangedlooptest.CallbackObserver{
+			OnStart: func(ctx context.Context, t time.Time) error {
+				incShouldBeCalled()
+				return nil
+			},
+			OnFork: func(ctx context.Context) (rangedloop.Partial, error) {
+				incShouldBeCalled()
+				return nil, nil
+			},
+			OnProcess: func(ctx context.Context, segments []segmentloop.Segment) error {
+				incShouldBeCalled()
+				return nil
+			},
+			OnJoin: func(ctx context.Context, partial rangedloop.Partial) error {
+				incShouldBeCalled()
+				return nil
+			},
+			OnFinish: func(ctx context.Context) error {
+				incShouldBeCalled()
+				return errors.New("Test OnFinish error")
+			},
+		},
+		&rangedlooptest.CallbackObserver{
+			OnStart: func(ctx context.Context, t time.Time) error {
+				incShouldBeCalled()
+				return nil
+			},
+			OnFork: func(ctx context.Context) (rangedloop.Partial, error) {
+				incShouldBeCalled()
+				return nil, nil
+			},
+			OnProcess: func(ctx context.Context, segments []segmentloop.Segment) error {
+				incShouldBeCalled()
+				return nil
+			},
+			OnJoin: func(ctx context.Context, partial rangedloop.Partial) error {
+				incShouldBeCalled()
+				return nil
+			},
+			OnFinish: func(ctx context.Context) error {
+				incShouldBeCalled()
+				return nil
+			},
+		},
+	}
+
+	loopService := rangedloop.NewService(
+		zaptest.NewLogger(t),
+		rangedloop.Config{
+			BatchSize:   batchSize,
+			Parallelism: parallelism,
+		},
+		&rangedlooptest.RangeSplitter{
+			Segments: segments,
+		},
+		observers,
+	)
+
+	observerDurations, err := loopService.RunOnce(testcontext.New(t))
+	require.NoError(t, err)
+	// only the 2 completed observers should have the duration reported
+	require.Len(t, observerDurations, len(observers))
+	require.Equal(t, int32(38), shouldBeCalled)
```

Comment:
nit: use EqualValues

also wondering how we can make 38 less magical

Reply:
I will split this out into 5 counters:

- numOnStartCalls
- numOnForkCalls
- numOnProcessCalls
- numOnJoinCalls
- numOnFinishCalls